File: local/app/game/tetris/piece.py
"""方块定义 & 旋转（SRS 墙踢系统）"""
import random
from constants import SHAPES, PIECE_NAMES, WALL_KICKS_JLSTZ, WALL_KICKS_I, COLS, ROWS


class Piece:
    """单个活动方块"""

    def __init__(self, name: str):
        self.name = name
        self.type = PIECE_NAMES.index(name)
        self.rotation = 0
        self.x = COLS // 2 - 2  # 从中间偏左生成
        self.y = -1              # 从顶部上方开始（避免立即碰撞）
        self.shape = SHAPES[name][0]

    def get_shape(self, rotation: int = None):
        """获取指定旋转的矩阵"""
        r = rotation if rotation is not None else self.rotation
        return SHAPES[self.name][r]

    def get_blocks(self) -> list[tuple[int, int]]:
        """返回方块占用的所有格子 (col, row)"""
        blocks = []
        for r in range(4):
            for c in range(4):
                if self.shape[r][c]:
                    blocks.append((self.x + c, self.y + r))
        return blocks

    def try_move(self, dx: int, dy: int, board: list[list[int]]) -> bool:
        """尝试移动并返回是否成功"""
        if self._can_place(self.x + dx, self.y + dy, self.rotation, board):
            self.x += dx
            self.y += dy
            return True
        return False

    def try_rotate(self, board: list[list[int]], clockwise: bool = True) -> bool:
        """SRS 旋转 + 墙踢"""
        from_rot = self.rotation
        to_rot = (from_rot + 1) % 4 if clockwise else (from_rot + 3) % 4

        kicks = WALL_KICKS_I if self.name == "I" else WALL_KICKS_JLSTZ
        offsets = kicks.get((from_rot, to_rot), [(0, 0)])

        for dx, dy in offsets:
            if self._can_place(self.x + dx, self.y - dy, to_rot, board):
                self.x += dx
                self.y -= dy
                self.rotation = to_rot
                self.shape = self.get_shape()
                return True
        return False

    def _can_place(self, px: int, py: int, rotation: int, board: list[list[int]]) -> bool:
        """检查方块在指定位置是否合法"""
        shape = self.get_shape(rotation)
        for r in range(4):
            for c in range(4):
                if shape[r][c]:
                    col = px + c
                    row = py + r
                    if col < 0 or col >= COLS or row >= ROWS:
                        return False
                    if row >= 0 and board[row][col]:
                        return False
        return True

    def is_above_board(self) -> bool:
        """检查是否有任何块在顶部上方（spawn 失败检测）"""
        for r in range(4):
            for c in range(4):
                if self.shape[r][c]:
                    if self.y + r < 0:
                        return True
        return False


class BagRandomizer:
    """7-Bag 随机系统"""

    def __init__(self):
        self.bag = []

    def next(self) -> str:
        if not self.bag:
            self.bag = list(PIECE_NAMES)
            random.shuffle(self.bag)
        return self.bag.pop()

File: local/app/game/tetris_cpp/src/piece.cpp
#include "piece.h"
#include <algorithm>

Piece::Piece(int t) : type(t), rotation(0), x(SPAWN_X), y(SPAWN_Y) {
    auto& sd = (*ALL_SHAPES[type])[rotation];
    width = sd.width;
    height = sd.height;
}

uint16_t Piece::mask() const {
    return (*ALL_SHAPES[type])[rotation].mask;
}

bool Piece::bit_at(uint16_t m, int row, int col) const {
    int bit = 15 - (row * 4 + col);
    return (m >> bit) & 1;
}

std::vector<Block> Piece::get_blocks() const {
    std::vector<Block> blocks;
    uint16_t m = mask();
    for (int r = 0; r < 4; ++r) {
        for (int c = 0; c < 4; ++c) {
            if (bit_at(m, r, c)) {
                blocks.push_back({x + c, y + r});
            }
        }
    }
    return blocks;
}

bool Piece::can_place(int px, int py, int rot, const int grid[ROWS][COLS]) const {
    uint16_t m = (*ALL_SHAPES[type])[rot].mask;
    for (int r = 0; r < 4; ++r) {
        for (int c = 0; c < 4; ++c) {
            if (!bit_at(m, r, c)) continue;
            int nr = py + r;
            int nc = px + c;
            // 允许顶部以上
            if (nc < 0 || nc >= COLS || nr >= ROWS) return false;
            if (nr >= 0 && grid[nr][nc] != 0) return false;
        }
    }
    return true;
}

bool Piece::is_above_board() const {
    for (auto& b : get_blocks()) {
        if (b.row < 0) return true;
    }
    return false;
}

bool Piece::try_move(int dx, int dy, const int grid[ROWS][COLS]) {
    if (can_place(x + dx, y + dy, rotation, grid)) {
        x += dx;
        y += dy;
        return true;
    }
    return false;
}

bool Piece::try_rotate(const int grid[ROWS][COLS], bool clockwise) {
    int from = rotation;
    int to;
    if (clockwise) {
        to = (rotation + 1) % 4;
    } else {
        to = (rotation + 3) % 4;
    }

    const KickEntry* kick;
    int key;
    if (clockwise) {
        key = from;
    } else {
        // 反向: to -> from 的顺时针旋转的反向 (取反偏移)
        key = to;
    }

    if (type == 0) {  // I piece
        kick = &WALL_KICKS_I[key];
    } else {
        kick = &WALL_KICKS_JLSTZ[key];
    }

    const auto& tests = kick->tests;
    for (int i = 0; i < 5; ++i) {
        int dx = tests[i].first;
        int dy = tests[i].second;

        // 顺时针: 正向偏移; 逆时针: 取反
        if (!clockwise) {
            dx = -dx;
            dy = -dy;
        }

        if (can_place(x + dx, y + dy, to, grid)) {
            x += dx;
            y += dy;
            rotation = to;
            auto& sd = (*ALL_SHAPES[type])[rotation];
            width = sd.width;
            height = sd.height;
            return true;
        }
    }
    return false;
}

bool Piece::in_bounds(int col, int row) const {
    return col >= 0 && col < COLS && row < ROWS;
}

// ---- BagRandomizer ----

int BagRandomizer::next() {
    if (bag.empty() || index >= (int)bag.size()) {
        refill();
    }
    return bag[index++];
}

void BagRandomizer::refill() {
    bag = {0, 1, 2, 3, 4, 5, 6};
    std::shuffle(bag.begin(), bag.end(), rng);
    index = 0;
}
